rust/chains/abacus-ethereum/src/multisig_ism.rs

```diff
@@ -0,0 +1,237 @@
+        Self {
+            contract: Arc::new(EthereumMultisigIsmInternal::new(
+                &locator.address,
+                provider.clone(),
+            )),
+            domain: locator.domain,
+            chain_name: locator.chain_name.to_owned(),
+            provider,
+            threshold_cache: Mutex::new(ExpiringCache::new(Duration::from_secs(60))),
+            validators_cache: Mutex::new(ExpiringCache::new(Duration::from_secs(60))),
+        }
+    }
+}
+
+impl<M> AbacusContract for EthereumMultisigIsm<M>
+where
+    M: Middleware + 'static,
+{
+    fn chain_name(&self) -> &str {
+        &self.chain_name
+    }
+
+    fn address(&self) -> H256 {
+        self.contract.address().into()
+    }
+}
+
+#[async_trait]
+impl<M> MultisigIsm for EthereumMultisigIsm<M>
+where
+    M: Middleware + 'static,
+{
+    /// Returns the metadata needed by the contract's verify function
+    async fn format_metadata(
+        &self,
+        checkpoint: &MultisigSignedCheckpoint,
+        proof: Proof,
+    ) -> Result<Vec<u8>, ChainCommunicationError> {
+        let threshold = self.threshold(checkpoint.checkpoint.mailbox_domain).await?;
+        let validators: Vec<H256> = self
+            .validators(checkpoint.checkpoint.mailbox_domain)
+            .await?
+            .iter()
+            .map(|&x| H256::from(x))
+            .collect();
+        let validator_tokens: Vec<Token> = validators
+            .iter()
+            .map(|x| Token::FixedBytes(x.to_fixed_bytes().into()))
+            .collect();
+        let proof_tokens: Vec<Token> = proof
+            .path
+            .iter()
+            .map(|x| Token::FixedBytes(x.to_fixed_bytes().into()))
+            .collect();
+        let prefix = ethers::abi::encode(&[
+            Token::FixedBytes(checkpoint.checkpoint.root.to_fixed_bytes().into()),
+            Token::Uint(U256::from(checkpoint.checkpoint.index)),
+            Token::FixedBytes(
+                checkpoint
+                    .checkpoint
+                    .mailbox_address
+                    .to_fixed_bytes()
+                    .into(),
+            ),
+            Token::FixedArray(proof_tokens),
+            Token::Uint(threshold),
+        ]);
+        let suffix = ethers::abi::encode(&[Token::FixedArray(validator_tokens)]);
+        // The ethers encoder likes to zero-pad non word-aligned byte arrays.
+        // Thus, we pack the signatures, which are not word-aligned, ourselves.
+        let signature_vecs: Vec<Vec<u8>> =
+            checkpoint.signatures.iter().map(|x| x.to_vec()).collect();
+        let signature_bytes = signature_vecs.concat();
+        let metadata = [prefix, signature_bytes, suffix].concat();
+        Ok(metadata)
+    }
+
+    #[tracing::instrument(err, skip(self))]
+    async fn threshold(&self, domain: u32) -> Result<U256, ChainCommunicationError> {
+        let mut cache = self.threshold_cache.lock().await;
```

Comment:
Noticing that the lock is held while we're making the call to get the threshold if there's a cache miss

Given that this code path is unlikely to have a lot of contention, I don't think it's a huge deal, but this is supposedly a bad practice. You can read about it here https://tokio.rs/tokio/tutorial/shared-state#holding-a-mutexguard-across-an-await

I'd like @mattiecnvr's input as someone who's more familiar with rust/tokio best practices, but I think we should consider:
1. Using an `std::sync::Mutex` -- everywhere I've read, people tend to advise against using the tokio one
2. not holding the lock when making the call to get the threshold
3. consider some code reuse between the `threshold` and `validators` fns, they seem pretty similar. Don't feel strongly here though

Reply:
> `std::sync::RwLockReadGuard<'_, ExpiringCache<u32, Vec<ethers::prelude::H160>>>` cannot be sent between threads safely
within `impl Future<Output = Result<Vec<ethers::prelude::H160>, ChainCommunicationError>>`, the trait `Send` is not implemented for `std::sync::RwLockReadGuard<'_, ExpiringCache<u32, Vec<ethers::prelude::H160>>>`

Using `tokio::sync::RwLock` works though

Reply:
> consider some code reuse between the threshold and validators fns, they seem pretty similar. Don't feel strongly here though

I'd advocate we do the quick and dirty thing here given my weak rust-fu, just given that a lot of this code is going to change anyways once we do full sovereign consensus support